src/pudl/analysis/allocate_net_gen.py

```diff
@@ -324,183 +509,437 @@ def stack_generators(
     return gens_stack
 
 
-def associate_generator_tables(gf, gen, gens):
-    """Associate the three tables needed to assign net gen to generators.
+def associate_generator_tables(
+    gens: pd.DataFrame, gf: pd.DataFrame, gen: pd.DataFrame, bf_by_gens: pd.DataFrame
+) -> pd.DataFrame:
+    """Associate the three tables needed to assign net gen and fuel to generators.
+
+    The ``generation_fuel_eia923`` table's data is reported at the
+    :py:const:`IDX_PM_ESC` granularity. Each generator has one ``prime_mover_code``, but
+    potentially several ``energy_source_code``s (reported in as primary source,
+    secondary source, etc.). Much of the data in ``generation_fuel_eia923`` not
+    associated with primary energy sources, so before merging the data with the
+    generators, we employ :func:``stack_generators``.
+
+    After the generator's energy source codes have been stacked, we group the other
+    data tables and merge them with the generators. We add suffixes to the data columns
+    to identify the source table.
```

Comment:
Group by what? Merge on what?

Reply:
I hesitated explaining here bc there are so many different groups and merges in this function.

basically we are getting sums/totals at different granularities to be used in the allocation function.